Sources/TSCBasic/FileSystem.swift

```diff
@@ -224,41 +234,78 @@ public extension FileSystem {
     func getFileInfo(_ path: AbsolutePath) throws -> FileInfo {
         fatalError("This file system currently doesn't support this method")
     }
+
+    func resolveUnicode(_ path: AbsolutePath) -> AbsolutePath {
+        return path
+    }
 }
 
 /// Concrete FileSystem implementation which communicates with the local file system.
 private class LocalFileSystem: FileSystem {
 
+    func resolveUnicode(_ path: AbsolutePath) -> AbsolutePath {
+        #if os(macOS)
+        // The macOS file system enforces NFD.
+        // It can find everything without any help.
+        return path
+        #else
```

Comment:
Hmm, this will tank the performance on non-macOS platforms. Can you check how corelibs and llvm::path handle these? I wonder if there's a API in corelibs that we can leverage here.

Reply:
I think `corelibs` is just as broken in this regard, since most uses here are immediately followed by calls to `FileManager`, whose failures clued me into this in the first place.

I don’t find anything relevant in corelibs when searching for “Unicode”, or when perusing the FileManager and URL sources. Maybe someone from that team would know better? @millenomi?

I find nothing related in `llvm::sys::path` or `llvm::sys::unicode` either.

Those lower levels may simply not care. The package manager is in an unusual situation, in that it looks up paths that it neither received from the file system nor placed there itself (and also in that it cannot defer this responsibility to a higher‐level client). This is the gap in which lookup becomes necessary. SwiftPM’s unique case is compounded by the fact that the same string literal in a manifest may need to match differing byte representations when checked out on different devices, so even telling users to be explicit with their literals (`"\u{...}\u{...}\u{...}"`) cannot work around it.

Reply:
The intent for corelibs is to mimic Darwin. On Darwin, normalization occurs at the filesystem access level.

These should likely be contributions to FileManager to supplement its .fileSystemRepresentation… method. FileManager should accept any path as input and normalize it internally.

Reply:
i.e. this is the wrong place; we should make FileManager better.

Reply:
Yeah, I agree that this should be solved at the `FileManager` level.

Reply:
> we should make FileManager better.

This is good to hear.

Unfortunately, my Linux machine is not large enough to develop `core-libs-foundation`. I can just barely manage to test SwiftPM because it works in isolation, without any of the other `swift-...` repositories. So that probably means someone else will have to be the one to sink it from here.

[`fileSystemRepresentation(withPath:)`](https://github.com/apple/swift-corelibs-foundation/blob/ea46535062a92f7500439a96bbd43b537fd8dd91/Foundation/FileManager.swift#L1020-L1025) looks like it intends to do the same thing, but at the moment it appears to basically just forward to [`CFStringGetFileSystemRepresentation`](https://opensource.apple.com/source/CF/CF-635.19/CFStringEncodings.c.auto.html), which in turn is currently does nothing outside of macOS. I don’t know if that is the right place for it to live. On macOS it basically only needs to decompose the string, but on other platforms it must actually look things up in the file system.

Reply:
I haven't checked but maybe it's possible to develop corelibs-foundation against as a standalone project if you install a toolchain on Linux?